File: fdr1d_matlab/ex_anova3.py

import numpy as np
from matplotlib import pyplot
import spm1d
import fdr1d



#(0) Load data:
dataset      = spm1d.data.uv1d.anova3.SPM1D_ANOVA3_2x2x2()
# dataset      = spm1d.data.uv1d.anova3.SPM1D_ANOVA3_2x3x4()
Y,A,B,C      = dataset.get_data()



#(1) Conduct ANOVA:
alpha        = 0.05
FF           = spm1d.stats.anova3(Y, A, B, C, equal_var=True)
FFi          = FF.inference(0.05)
fstars_fdr   = [fdr1d.inference(F, alpha=0.05)  for F in FF]
print( FFi )



#(2) Plot results:
pyplot.close('all')
FFi.plot(plot_threshold_label=True, plot_p_values=True, autoset_ylim=True)

for ax,z in zip(pyplot.gcf().axes, fstars_fdr):
	if z is not None:
		ax.axhline(z, color='b', linestyle='--')
		ax.text(50, z-0.3, r'$t^*$ (FDR) = %.3f' %z, color='b', size=8)

pyplot.show()

File: fdr1d_matlab/ex_anova3.m
    

clear;  clc



%(0) Load data:
dataset    = spm1d.data.uv1d.anova3.SPM1D_ANOVA3_2x2x2();
% dataset    = spm1d.data.uv1d.anova3.SPM1D_ANOVA3_2x3x4();
[Y,A,B,C]  = deal(dataset.Y, dataset.A, dataset.B, dataset.C);


%(1) Conduct SPM analysis:
alpha      = 0.05;
two_tailed = false;
spmlist    = spm1d.stats.anova3(Y, A, B, C);
spmilist   = spmlist.inference(0.05);
fstar      = inference_fdr(spmlist, alpha, two_tailed);
disp_summ(spmilist)


%(2) Plot:
close all
spmilist.plot('plot_threshold_label',false, 'plot_p_values',true, 'autoset_ylim',true);

% %MANUALLY construct the same plot:
for k = 1:spmilist.nEffects
    subplot(3,3,k)
    spmi = spmilist(k);
    spmi.plot();
    title( spmi.effect )
    line( xlim, [fstar(k) fstar(k)], 'color', 'blue', 'linestyle', '--');
    text( mean(xlim), fstar(k), sprintf('F* (FDR) = %.3f', fstar(k)), 'color', 'blue')
end